src/Common/LRUCache.h

```diff
@@ -286,6 +317,15 @@ class LRUCache
 
         if (inserted)
         {
+            auto value_weight = mapped ? weight_function(*mapped) : 0;
+            // move removeOverflow() ahead here. In default, the final result is the same as the old implementation
+            if (!removeOverflow(value_weight))
+            {
+                // cannot find enough space to put in the new value
+                cells.erase(it);
+                return false;
+            }
+
```

Comment:
May be we can avoid this part by making sure that `canRelease` is true for value which we trying to add right now?

Reply:
we need to guarantee that the element must be in the cache once set() returns true. If we remove this part, it may cause the new value be evicted immediately

Reply:
Ok, agree.

Reply:
I think that after all you do not need this. Because look, you anyway need to call `set` under mutex in singleton class. So in this case you can first check if there is available place (with `maxSize` and `wight()` methods) and then do your `set`.

Reply:
Yes, it is another choice